src/protocol/boolean/bit_decomposition.rs

```diff
@@ -38,49 +40,170 @@ impl BitDecomposition {
         S: LinearSecretSharing<F> + BasicProtocols<C, F>,
         C: Context + RandomBits<F, Share = S>,
     {
-        // step 1 in the paper is just describing the input, `[a]_p` where `a ∈ F_p`
-
-        // Step 2. Generate random bitwise shares
+        // Step 1. Generate random bitwise shares [r]_B and linear share [r]_p
         let r = rbg.generate(record_id).await?;
 
-        // Step 3, 4. Reveal c = [a - b]_p
-        let c = (a_p.clone() - &r.b_p)
+        // Step 2: Reveal c = [a - r]_p
+        let c: F = (a_p.clone() - &r.b_p)
             .reveal(ctx.narrow(&Step::RevealAMinusB), record_id)
             .await?;
 
-        // Step 5. Add back [b] bitwise. [d]_B = BitwiseSum(c, [b]_B) where d ∈ Z
-        //
-        // `BitwiseSum` outputs one more bit than its input, so [d]_B is (el + 1)-bit long.
-        let d_b = add_constant(ctx.narrow(&Step::AddBtoC), record_id, &r.b_b, c.as_u128()).await?;
+        // Step 2.1: Edge case, if r is coincidentally a bit decomposition of [a]_p, we're done.
+        if c == F::ZERO {
+            return Ok(r.b_b);
+        }
 
-        // Step 6. q = d >=? p (note: the paper uses p <? d, which is incorrect)
-        let q_p = BitwiseLessThanPrime::greater_than_or_equal_to_prime(
-            ctx.narrow(&Step::IsPLessThanD),
+        // Step 3. Compute [q]_p = 1 - [r <_B p - c ]_p. q is 1 iff r + c >= p in the integers.
+        let p_minus_c: u128 = F::PRIME.into() - c.as_u128();
+        let q_p = bitwise_less_than_constant(
+            ctx.narrow(&Step::IsRLessThanPMinusC),
             record_id,
-            &d_b,
+            &r.b_b,
+            p_minus_c.into(),
         )
         .await?;
 
-        // Step 7. a bitwise scalar value `f_B = bits(2^el - p)`
+        // Step 4 has a lot going on. We'd going to break it down into substeps.
+        // Step 4.1. Make a bitwise scalar value of f = 2^el + c - p.
+        // let el = usize::try_from(u128::BITS - F::PRIME.into().leading_zeros()).unwrap();
         let el = u128::BITS - F::PRIME.into().leading_zeros();
-        let x = (1 << el) - F::PRIME.into();
+        let two_exp_el = u128::pow(2, el);
+        let _f_int: u128 = two_exp_el + c.as_u128() - F::PRIME.into();
+        debug_assert!(_f_int < u64::max_value().into());
+        let f_int: u64 = _f_int as u64;
+
+        // Step 4.2. Compute [g]_B = (f_i - c_i) [q]_p + c_i
+        let mut g_bin: Vec<G> = Vec::with_capacity(usize::try_from(el).unwrap());
```

Comment:
you probably shouldn't keep these values, rather compute that on the fly because it is very cheap and requires only two 8 byte values to be around

you can still define an Enum to represent those logical values and use it, just compute it on the fly

Reply:
yep good call. I was thinking the same thing. the two loops can become one